Fixed #33012 -- Added Redis cache backend. (#14437)
django/django: docs/topics/cache.txt

    some problems and seems to be unmaintained. Use ``PyMemcacheCache`` or
    ``PyLibMCCache`` instead.

.. _redis:

Redis
-----

Redis__ is an in-memory database that can be used for caching. To begin you'll
need a Redis server running either locally or on a remote machine.

__ https://redis.io/

After setting up the Redis server, you'll need to install Python bindings for
Redis. `redis-py`_ is the binding supported natively by Django.

.. _`redis-py`: https://pypi.org/project/redis/

To use Redis as your cache backend with Django:

> * Set :setting:`BACKEND <CACHES-BACKEND>` to
>   ``django.core.cache.backends.redis.RedisCache``

Reviewer: ```suggestion
* Set :setting:`BACKEND <CACHES-BACKEND>` to
  ``django.core.cache.backends.redis.RedisCache``.
```

> * Set :setting:`LOCATION <CACHES-LOCATION>` to the URL pointing to your Redis 
>   instance, using the appropriate scheme. See the example below, and the 
>   ``redis-py`` docs for `details on the available schemes 
>   <https://redis-py.readthedocs.io/en/stable/#redis.ConnectionPool.from_url>`_. 

Reviewer: Remove trailing spaces:
```suggestion
* Set :setting:`LOCATION <CACHES-LOCATION>` to the URL pointing to your Redis
  instance, using the appropriate scheme. See the example below, and the
  ``redis-py`` docs for `details on the available schemes
  <https://redis-py.readthedocs.io/en/stable/#redis.ConnectionPool.from_url>`_.
```

For example, if Redis is running on localhost (127.0.0.1) port 6379::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379',
        }
    }

Often Redis servers are protected with authentication. In order to supply a
username and password, add them in the ``LOCATION`` along with the URL::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',

>             'LOCATION': 'rediss://username:password@127.0.0.1:6379',

Reviewer: ```suggestion
            'LOCATION': 'redis://username:password@127.0.0.1:6379',
```
....ssss :snake:

Reviewer 2: Actually, this isn't incorrect. `rediss://` is for Redis-over-TLS. See the following:

- https://github.com/jazzband/django-redis#configure-as-cache-backend
- https://www.iana.org/assignments/uri-schemes/prov/rediss

But it does look like a typo...

Author: No I actually did this on purpose 😛 
Since we had examples of `redis://` and `unix://` for the `LOCATION`, I though of using `rediss://` for this one!

Reviewer 2: > No I actually did this on purpose 

I get that. I meant that the choice of `rediss` as the scheme name, while it follows `http` → `https`, just makes it look like a typo when it isn't.

Reviewer: I don't think that we need examples for all available options. Switching between `redis` and `rediss` without any comment, only to put all available schemas in the docs is confusing. I'd use `redis` instead.

        }
    }

> If you have multiple Redis servers set up in replication mode, you can specify
> the servers either as a semicolon or comma delimited string, or as a list.
> While using multiple servers, all the writes are performed on the lead server.
> Read operations are performed on one of the replicas chosen at random. The first
> server address must be the leader server and the other server addresses must be
> the followers (replicas of leader)::

Reviewer: `the` is missing :thinking:. Also, this is too long and a bit confusing, maybe:
```suggestion
If you have multiple Redis servers set up in the replication mode, you can
specify the servers either as a semicolon or comma delimited string, or as a
list. While using multiple servers, write operations are performed on the first
server (leader). Read operations are performed on the other servers (replicas)
chosen at random::
```

  for each one. The ``OPTIONS`` ``db`` argument can be used in such cases for 
  clarity.

* ``pool_class``: Connection pools are a set of connections which are used as 
  required. ``pool_class`` allows you to set a custom connection pool class.
  Defaults to ``redis.ConnectionPool``. See the ``redis-py`` `connection pool
  docs`_ for more details .
  
  .. _`connection pool docs`: https://github.com/andymccurdy/redis-py#connection-pools

* ``parser_class``: Parser classes control on how the responses from the
  Redis server are parsed. By default ``hiredis`` will be used if installed
  otherwise the ``PythonParser`` will be used. ``Hiredis`` is written in C and
  therefore can give up to 10x gains over the ``PythonParser``.
  See the ``redis-py`` docs for more about `parsers`_.

  .. _`parsers`: https://github.com/andymccurdy/redis-py#parsers

In the following example, we set the ``parser_class`` to
``redis.connection.HiredisParser``, the ``db`` to 12 and the ``pool_class`` to
``redis.BlockingConnectionPool``::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',

>             'LOCATION': 'unix:///var/run/redis/redis.sock',

Reviewer: Using a single slash would also be consistent with Memcached examples.

            'OPTIONS': {
                'pool_class': 'redis.BlockingConnectionPool',
                'parser_class': 'redis.connection.HiredisParser',
                'db': 12,
            }
        }

>     }

Reviewer: Extra options for other backends are described shortly in the [_"Cache arguments"_](https://docs.djangoproject.com/en/dev/topics/cache/#cache-arguments) section. IMO we don't need a detailed description and many examples. Maybe:
```
Here's an example configuration for a ``redis`` based backend that chooses the
10th database (by default Redis ships with 16 logical databases), sets a custom
`connection pool class`_ (``redis.ConnectionPool`` is used by default), and ...

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379',
            'OPTIONS': {
                'pool_class': 'redis.BlockingConnectionPool',
                'parser_class': 'redis.connection.HiredisParser',
                'db': '10',
            }
        }   
    }   
    
.. _`connection pool docs`: https://github.com/andymccurdy/redis-py#connection-pools
.. _`parsers`: https://github.com/andymccurdy/redis-py#parsers
```

Author: I think a little context with some examples will make things easier for the user. Moreover in the future, we'll be added more options for `username` and `password` too. Also, since there are warnings (in a multi-server setup) and various ways of setting `db` etc, having a separate section would give better organization and clear understanding of the arguments in my opinion.

Author: However in the long run, if we write reference documentation, these can move away from here and we can keep a single example here. Read [here](https://github.com/django/django/pull/14437/files#r689823015)
Thoughts on this?

    some problems and seems to be unmaintained. Use ``PyMemcacheCache`` or
    ``PyLibMCCache`` instead.

> .. _redis:

Author: Missed that, thank you!

Redis
-----

> Redis__ is the most popular cache backend used with Django. It is an in-memory

Author: Initially I added the survey but @ngnpope suggested otherwise.
https://github.com/django/django/pull/14437#discussion_r658860268

database which stores key-value pairs. Redis server can be configured to run as
a daemon either on the local machine or on a remote instance.

__ https://redis.io/

> After setting up the redis server, you'll need to install redis bindings for

Author: I thought about this but I took inference from the memcached documentation and there memcached is not capitalized. Hence I stuck with the lowercase. Let me know if I should capitalize it! 
CC: @carltongibson

> After setting up the redis server, you'll need to install redis bindings for

Reviewer 3: Let's ask @abrookins. Andrew, should we be capitalising Redis?

.. _`redis-py`: https://pypi.org/project/redis/

To use redis as your cache backend with Django:

* Set :setting:`BACKEND <CACHES-BACKEND>` to
  ``django.core.cache.backends.redis.RedisCache``

* Set :setting:`LOCATION <CACHES-LOCATION>` to ``ip:port`` values, where ``ip``
  is the IP address of the redis server and ``port`` is the port on which redis
  is running, or to a ``unix:path`` value, where ``path`` is the path to a redis
  Unix socket file.

For example, if redis is running on localhost (127.0.0.1) port 6379::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379',
        }
    }

Often redis servers are protected with authentication. If order to supply a
username and password, we add them in the ``LOCATION`` along with the URL::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'rediss://username:password@127.0.0.1:6379',
        }
    }

If you have multiple redis servers setup in replication mode, you can specify
the servers in a list. While using multiple servers, all the writes are

> performed on the leader server. The read operation is performed on either of the

Author: Oh yes, this makes more sense.

replicas chosen at random. The first server address must be the leader server
and the other server addresses must be the followers (replicas of leader)::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': [
                "redis://127.0.0.1:6379", # leader
                "redis://127.0.0.1:6378", # read-replica 1
                "redis://127.0.0.1:6377", # read-replica 2
            ],
        }
    }

Along with the standard :ref:`cache_arguments`, the redis caching backend also

> accepts the following additional arguments:

Reviewer 3: Maybe options rather than arguments?

* ``db``: By default, redis ships with 16 logical databases. A database of your
  choice can be selected either by setting it in the ``LOCATION`` as querystring
  or path argument, or by setting it in the ``OPTIONS``. The default choice for
  the database is 0. We can set the database in the following manner::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',

>             'LOCATION': 'redis://127.0.0.1:6379?db=10', # option 1

Reviewer 3: This looks too much like a full example. However, I think you're saying use on of these three options. I wonder if there's a better way of presenting this.

Author: So initially I had three separate examples for the three separate example but the seemed to be too lengthy. Hence I reverted to condensing them into one. Since all three options are not specified in the same manner, I went with the elaborate example. Let me know what you feel!

            'LOCATION': 'redis://127.0.0.1:6379/10', # option 2
            'OPTIONS': {
                'db': '10' # option 3
            }
        }
    }

  Note that if multiple values of database are given, the first value
  encountered according to the above order would be considered.

  Also note that in a multi redis server setup, specifying different databases
  in the URL would not work as the redis replication works only on the same
  database number across the servers, hence the same database must be selected.
  The ``db`` argument in ``OPTIONS`` can be used in such cases to avoid
  confusion.

> * ``pool_class``: Connection pools create a set of connection which are

Reviewer 3: ```suggestion
* ``pool_class``: Connection pools create a set of connections which are
```

  retrieved as and when required. You can set it to the connection pool of your
  choice or even develop a custom connection pool. This defaults to
  ``redis.ConnectionPool``. Read more about `connection-pools`_.

  .. _`connection-pools`: https://github.com/andymccurdy/redis-py#connection-pools

* ``parser_class``: Parser classes can control on how the responses from the Redis
  server are parsed. Few parsers like ``hiredis`` can give around 10x speed up. The

>   default choice is ``redis.connection.HiredisParser``, if hiredis is installed

Reviewer 3: Maybe?
```suggestion
  default choice is ``redis.connection.HiredisParser``, if ``hiredis`` is installed
```

> * ``parser_class``: Parser classes can control on how the responses from the Redis
>   server are parsed. Few parsers like ``hiredis`` can give around 10x speed up. The
>   default choice is ``redis.connection.HiredisParser``, if hiredis is installed
>   else it falls back on using the ``redis.connection.PythonParser``. Read more
>   about `parsers`_.

Reviewer 3: I suspect wrapping here isn't quite right. 

```suggestion
* ``parser_class``: Parser classes control how responses from the Redis
  server are parsed. By default ``hiredis`` will be used if installed otherwise 
  the ``PythonParser`` will be used. ``Hiredis`` is written in C and therefore can give
  up to 10x gains over PythonParser.
 Read more about `parsers`_.
```

  .. _`parsers`: https://github.com/andymccurdy/redis-py#parsers

>   If the following example, we  set the ``parser_class`` to

Reviewer 3: I'm not sure, but if above we say this batch here are all options, and given an example of what a `db` option looks like. Do we need this example as well? 🤷 

```suggestion
  In the following example, we set the ``parser_class`` to
```

  the database is 0. We can set the database in the following manner::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379',
            'OPTIONS': {
                'db': '10'
            }
        }
    }

  Read more about how `database selection`_.

  .. _`database selection`: https://redis-py.readthedocs.io/en/stable/#redis.ConnectionPool.from_url

  Also note that in a multi redis server setup, specifying different databases
  in the URL would not work as the redis replication works only on the same
  database number across the servers, hence the same database must be selected.
  The ``db`` argument in ``OPTIONS`` can be used in such cases to avoid
  confusion.

* ``pool_class``: Connection pools create a set of connections which are
  retrieved as and when required. You can set it to the connection pool of your
  choice or even develop a custom connection pool. This defaults to
  ``redis.ConnectionPool``. Read more about `connection-pools`_.

  .. _`connection-pools`: https://github.com/andymccurdy/redis-py#connection-pools

> * ``parser_class``: Parser classes control on how the responses from the
>   Redis server are parsed. By default ``hiredis`` will be used if installed
>   otherwise the ``PythonParser`` will be used. ``Hiredis`` is written in C and
>   therefore can give up to 10x gains over PythonParser.
>   Read more about `parsers`_.
> 
>   .. _`parsers`: https://github.com/andymccurdy/redis-py#parsers

Reviewer 4: ```suggestion
* ``parser_class``: Parser classes control on how the responses from the
  Redis server are parsed. By default ``hiredis`` will be used if installed
  otherwise the ``PythonParser`` will be used. ``Hiredis`` is written in C and
  therefore can give up to 10x gains over the ``PythonParser``.
  See the ``redis-py`` docs for more about `parsers`_.

  .. _`parsers`: https://github.com/andymccurdy/redis-py#parsers
```

  In the following example, we set the ``parser_class`` to
  ``redis.connection.HiredisParser`` and the ``pool_class`` to
  ``redis.BlockingConnectionPool``::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'unix:/var/run/redis/redis.sock',
            'OPTIONS': {
                'pool_class': 'redis.BlockingConnectionPool'
                'parser_class': 'redis.connection.HiredisParser'
            }
        }

>     }

Reviewer 4: Not sure about this example... 

* It could be outdented, since it's not part of the bullet-list. 
* If you add `db` then it's showing all the options listed. 
* Could mention that it's using the `unix://` Unix Domain Socket scheme. (Please check the URL. I think you need three `/` at the beginning, two for the scheme and one for the `/` root path. 🤔)

> 

Reviewer 4: We only need one blank line here.

    some problems and seems to be unmaintained. Use ``PyMemcacheCache`` or
    ``PyLibMCCache`` instead.

.. redis:

Redis
---------

Redis is the most used caching backend according to the developers survey.
...

> ...

Author: Yes I haven't really thought this through. Just wanted to add a placeholder for now.
Although, what all shall be mention here? I see in the Memcached section, there are few pro and performance benifits of Memcached mentioned, followed by the installation examples. Do we want something similar for Redis or we can skip that?

        }
    }

``Redis cache arguments``
~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Author: Yes I'll fix the wrapping issues. 

I was thinking of linking it to the redis-py documentation. However, there isn't must documentation [here](https://redis-py.readthedocs.io/en/stable/)

We can use their README.md for the descriptions. 
[Connection Pools](https://github.com/andymccurdy/redis-py#connection-pools)
[Parsers](https://github.com/andymccurdy/redis-py#connection-pools)

We can condense the examples into a single one with all the arguements. Working on it

The redis caching backend accepts the following additional arguments:

* ``pool_class``: Connection pools create a set of connection which are retrieved
  as and when required. You can set it to the connection pool of your choice or
  even develop a custom connection pool. This defaults to ``redis.ConnectionPool``::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379',
            'OPTIONS': {
                'pool_class': 'redis.BlockingConnectionPool'
            }
        }
    }

* ``db``: By default, Redis ships with 16 logical databases. A db of your choice can
  be selected either by mentioning it in the ``LOCATION`` as querystring or path
  argument, or set it in the ``OPTIONS``. The default choice for this is 0.
  We can set the db in the following manner::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'redis://127.0.0.1:6379?db=10', # option 1
            'LOCATION': 'redis://127.0.0.1:6379/10', # option 2
            'OPTIONS': {

>                 'db': '10' # option 3

Author: option 3 would be useful if we have multiple servers, then the user need not set the db in each URL rather, just specify it in the options.

  server are parsed. Few parsers like ``hiredis`` can give around 10x speed up. The
  default choice is ``redis.connection.HiredisParser``, if hiredis is installed
  else it falls back on using the ``redis.connection.PythonParser``::

    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.redis.RedisCache',
            'LOCATION': 'unix:/var/run/redis/redis.sock',
            'OPTIONS': {
                'parser_class': 'redis.connection.HiredisParser'
            }
        }
    }

* ``pickle_protocol``: Redis is capable of storing strings in the form of key-value
  pairs. Thus we need to serialize the data before storing it and deserialize it
  while retrieving it. Python's pickle module is used for this task. We can select
  the pickle protocol used for the serialization. This defaults to -1 (highest).::

    CACHES = {
            'default': {
                'BACKEND': 'django.core.cache.backends.redis.RedisCache',
                'LOCATION': 'rediss://127.0.0.1:6379',
                'OPTIONS': {
                    'pickle_protocol': 3 # highest protocol
                }
            }

>         }

Author: In that case, we can remove the `PickleSerializer` for good and simply use the `pickle.dumps` and `pickle.loads` with `protocol=HIGHEST_PROTOCOL` just like in few of the other cache backends.

>         }

Reviewer 2: So `OPTIONS` is passed as the arguments to the client, e.g. to `pymemache.HashClient()` for `PyMemcacheCache`:

https://github.com/django/django/blob/54a30a7a00fea6c5e3702282ade6e0238e06de3b/django/core/cache/backends/memcached.py#L40

We generally don't need to document all of those options that can be passed to the underlying client.

However... Things are a little bit different here. We are not using a third-party library to provide the client class which is what we are doing for the memcached backends, but writing our own. As such, none of the options for our client class are documented anywhere. Having said that, I'm not sure we need to document it at this stage. We may decide to do so later if we expose additional functionality.

I think that it still makes sense to keep `PickleSerializer` and not pass in `pickle_protocol` to `OPTIONS`. We should then do the same as for `PyMemcacheCache` and pass some default options:

https://github.com/django/django/blob/54a30a7a00fea6c5e3702282ade6e0238e06de3b/django/core/cache/backends/memcached.py#L237-L242

(Note that `serde` is the "serializer/deserializer" instance.)

 So perhaps we should pass the serializer in a similar way and change `RedisCache.__init__()` to have:
 
```python
        self._options = {'serializer': PickleSerializer(), **params.get('OPTIONS', {})}
```

This would then allow the following configuration:

```python
                'OPTIONS': {
                    'serializer': PickleSerializer(pickle.DEFAULT_PROCOL),  # Use default protocol instead of highest.
                }
```

Or equally a custom serializer, e.g. using `msgpack`, could be provided.

In summary, let's keep it flexible but avoid documenting everything for now. Once documented as public API we can't refine it as easily. Get the basic thing working and merged first and then extend as required.

>         }

Author: Yes this sounds like a better approach. Making the changes now and leaving the documentation for this for now
